### ISSUE TITLE
ble_gap_adv_start() ignores own_addr_type parameter (IDFGH-10724)

### ISSUE THREAD

**Issue by zacwbond** [espressif/esp-idf#11944]
### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### IDF version.

v4.4.3-3-g67b1d7c985

### Operating System used.

Linux

### How did you build your project?

Command line with CMake

### If you are using Windows, please specify command line type.

None

### Development Kit.

N/A

### Power Supply used.

External 3.3V

### What is the expected behavior?

I am trying to use a random static address for advertising (i.e., BLE spec *6.B.1.3.2.1 Static Device Address*).  I expect that once the address is configured, calling `ble_gap_adv_start()` with the first parameter parameter `own_addr_type` set to `BLE_OWN_ADDR_RANDOM` would cause the radio to begin advertising with a random address.

### What is the actual behavior?

There is no difference between using `BLE_OWN_ADDR_RANDOM` or `BLE_OWN_ADDR_PUBLIC`.  In both cases, the radio advertises using the public BLE device address.  To check the advertised address, I used nRF Connect on an Android phone.

### Steps to reproduce.

This code is running on an ESP32 installed on custom hardware.  Here's portions of my configuration code.  Maybe someone can show me if I missed a step somewhere.  

This is the stack initialization:
```
ESP_ERROR_CHECK(esp_nimble_hci_and_controller_init());
nimble_port_init();

// Initialize Nimble host callbacks.
ble_hs_cfg.reset_cb         = OnReset;
ble_hs_cfg.sync_cb          = OnSync;
ble_hs_cfg.store_status_arg = this;
ble_hs_cfg.store_status_cb = [](struct ble_store_status_event *event, void *arg)
{ return static_cast<EspBluetoothDriver *>(arg)->StoreStatus(event); };
ble_hs_cfg.sm_io_cap = BLE_SM_IO_CAP_NO_IO;

// Enable secure bonding.
ble_hs_cfg.sm_bonding = 1;
ble_hs_cfg.sm_mitm    = 0;
ble_hs_cfg.sm_sc      = 1;

// Use the same key mask for both.
ble_hs_cfg.sm_our_key_dist = 0x3;
ble_hs_cfg.sm_our_key_dist = 0x3;

ble_hs_cfg.sm_their_key_dist = ble_hs_cfg.sm_our_key_dist;

ble_svc_gap_device_name_set("default");

ble_store_config_init();

nimble_port_freertos_init(HostTask);
```

Here's the sync callback where the random address is generated and set.  It prints both the random and public addresses.
```
ESP_LOGE(BLE_DRIVER_TAG, "Sync detected.");

int rc;
ble_addr_t randAddress{};

if (rc = ble_hs_id_gen_rnd(0, &randAddress); rc != 0)
{
    ESP_LOGE(BLE_DRIVER_TAG, "Random address gen failed. Error %d", rc);
    return;
}

if (rc = ble_hs_id_set_rnd(randAddress.val); rc != 0)
{
    ESP_LOGE(BLE_DRIVER_TAG, "Random address failed. Error %d", rc);
    return;
}

if (rc = ble_hs_util_ensure_addr(0); rc != 0)
{
    ESP_LOGE(BLE_DRIVER_TAG, "Address config failed. Error %d", rc);
    return;
}

ble_addr_t ownAddress{};

// Print out our current random address.
ownAddress.type = BLE_OWN_ADDR_RANDOM;
if (rc = ble_hs_id_copy_addr(BLE_OWN_ADDR_RANDOM, ownAddress.val, nullptr), rc)
{
    ESP_LOGE(BLE_DRIVER_TAG, "Unable to load random address. Error %d", rc);
}

AddressString addressString{};
AddressToString(addressString, ownAddress);

ESP_LOGE(BLE_DRIVER_TAG, "BLE random address: %s", addressString);

// Print out our current public address.
ownAddress.type = BLE_OWN_ADDR_PUBLIC;
if (rc = ble_hs_id_copy_addr(BLE_OWN_ADDR_PUBLIC, ownAddress.val, nullptr), rc)
{
    ESP_LOGE(BLE_DRIVER_TAG, "Unable to load public address. Error %d", rc);
}

AddressToString(addressString, ownAddress);

ESP_LOGE(BLE_DRIVER_TAG, "BLE public address: %s", addressString);

ble_att_set_preferred_mtu(BafBle::MAX_MTU);
```

Here's how the advertisement is started:
```
ble_gap_adv_set_fields(&mAdvertisingData);
ble_gap_adv_rsp_set_fields(&mScanResponseData);

int error{ble_gap_adv_start(BLE_OWN_ADDR_RANDOM,
                                NULL,
                                BLE_HS_FOREVER,
                                &mAdvertisingParams,
                                GattCallback,
                                this)};
```

I didn't include the advertising or scan response parameters because they shouldn't matter and are very long.  

### Debug Logs.

This is the output of the sync function:

```plain
E (1053) Ble Driver: BLE random address: E4:78:3C:33:3F:F3
E (1054) Ble Driver: BLE public address: A8:03:2A:76:2A:DE
```

As you can see, the MSb of the random address are `0b11`, indicating this is a static random address.

Here's the output when I call `ble_gap_adv_start()`:

```
I (2722) BLE Peripheral: Start advertising
I (2722) NimBLE: GAP procedure initiated: advertise;
I (2722) NimBLE: disc_mode=2
I (2723) NimBLE:  adv_channel_map=7 own_addr_type=1 adv_filter_policy=0 adv_itvl_min=32 adv_itvl_max=64
I (2724) NimBLE:
```

As you can see, `own_addr_type` is set correctly here.  But when I pull out nRF Connect, I can see my product advertising with the address `A8:03:2A:76:2A:DE`.  

For what it's worth, at one point I turned on privacy mode and in that case I *do* get a RPA address when I start advertising.  But I need the static random address, and AFAIK I should not need privacy mode to do it.

### More Information.

This code is a change to an existing product to enable random static addresses because we are implementing Matter support, which requires the use of random static addresses.

I'd appreciate any corrections to the code I posted, or if there's a working example out there somewhere, feel free to direct me to it.  There's lots of issues related to RPA addresses, but not much around random static addresses.

**Comment by IshaESP** [comment]
Hi @zacwbond ,
For random address, ble_hs_util_ensure_addr(0) should be ble_hs_util_ensure_addr(1). 1 should be passed a parameter for preferring random address.
You can refer to bleprph example's code which is under `CONFIG_EXAMPLE_RANDOM_ADDR` this flag.

**Comment by zacwbond** [comment]
@IshaESP 

Thanks for the response.  In my case, I don't think ble_hs_util_ensure_addr() does much because I already called ble_hs_id_set_rnd() myself.  

But in further testing, I have found that there is an additional step required to reproduce the problem.  If I have *only* an advertiser, then it does use the random address consistently.  But my product is scanning and advertising at the same time, and in particular, we call:

`ble_gap_disc(BLE_OWN_ADDR_PUBLIC, SCAN_DURATION, &mScanParameters, GapEventCallback, this)` to start scanning.  As soon as this is called, then our ongoing advertisement switches to using the public address instead of the random address.  I can switch the parameter of `ble_gap_disc()` to BLE_OWN_ADDR_RANDOM and the problem goes away.  The advertisement continues to use the random address.

But there's another problem -- Our product is scanning because it wants to connect to bonded BLE remote controls.  In this case, we must use the public address in the connect call, as in `ble_gap_connect(BLE_OWN_ADDR_PUBLIC, &event->disc.addr,  CONNECTION_TIMEOUT_MS, NULL, GapEventCallback, this)`.    This also causes the advertisement to switch to the public address instead of the random address.  As soon as the connection succeeds or fails and scanning is restarted, the advertisement switches back to the random address.

In summary, it seems that the address used while advertising can change if `ble_gap_disc()` or `ble_gap_connect()` if either method is called with a different address type while an advertisement is ongoing.

**Comment by esp-zhp** [comment]
@zacwbond has the problem been resolved on your side?
Could you please let us know if the issue still exists when using the latest ESP-IDF branch?
